### DISCUSSION TITLE
Ability to access reverse relations within entry query in GQL

### DISCUSSION THREAD

**Discussion by andrewfairlie** [craftcms/cms#8954] [Ideas]
It would be really useful if when querying entries in GraphQL you could get _reverse_ related elements to that entry, which would avoid you having to do multiple GraphQL calls of "get the entry ID, and then run another GraphQL query with `relatedTo()` because this can quickly add up and become clunky. 

## Original comment from other thread

Just a thought, and apologies if this was covered in other thoughts above and I misunderstood, would a good syntax for this be something like `relatedEntries`, `relatedAssets` etc. which behave like a normal `entries` / `asset` loop but with the `relatedTo` pre-populated with the current entry. So you could do things like this very weird, but hopefully useful example...?

```
entry(section:"offices",slug:"london"){
   title,
   id,
   ... on offices_offices_entry {
      peopleInThisOffice: relatedEntries(section:"people"){
        title
      }
      facilitiesInThisOffice: relatedEntries(section:"facilities"){
        title
      }
      imagesTaggedInThisOffice: relatedAssets(source:"images"){
        title
      }
   }
}
```

_Originally posted by @andrewfairlie in https://github.com/craftcms/cms/issues/5143#issuecomment-721702205_

**Comment by andris-sevcenko** [comment]
This has been added for Craft 3.6

**Comment by andris-sevcenko** [comment]
Well, in its own branch for now.

**Comment by brandonkelly** [comment]
Merged now. Note that it’s `relatedToEntries`, etc., not `relatedEntries`.

**Comment by brandonkelly** [comment]
Craft 3.6 RC2 is out now with these new GraphQL arguments.

**Comment by brandonkelly** [comment]
Craft 3.6 has now been officially released ✨

**Comment by membla** [comment]
Just had a conversation on Discord with @andris-sevcenko on this – the actual issue at hand here is not resolved and even if it may not be feasible to do it in the GraphQL API without changes to the Element Query API or Element itself I think the issue should be reopened because the use case is very valid (and well explained in the original post above).

**Comment by brandonkelly** [comment]
Whoops, you’re right @membla. Guess we were on a different brainwave. We’re considering how to best address this.

**Comment by wbrowar** [comment]
I just wanted to throw some extra support behind this FR. I keep running into places where I do one query to get a URI, just to pass that single URI into a second `relatedToEntries` query. I've tried pre-loading all of my related content, but that means pulling in extra data and then doing the relations in JavaScript, which defeats the benefit of having a good relation system in Craft.

Having custom directives like the ones noted above would be a big boost to SPA performance and reduce static generated build times, so I'm really hoping to see this get added. Happy to help with testing or coming up with the approach to this.

**Comment by ctigelaar** [comment]
Since a milestone 4.0 tag was added back to the original issue in Feb 2021, I was really looking forward to the release of Craft 4 that would address this problem. I've checked out Craft 4 since it's release, but I still can't query the data of reverse relations unfortunately, or I'm missing something. 

What are the plans with this issue? The GraphQL interface is really lacking functionality here. I already had several cases where I needed to skip functionality because I simply can't get the related data based on the selection in the CMS. Even matching two sets of data is not possible, since no output of reverse relations is available at all. In the end what is the point of creating relations in the CMS if we can't use them in our applications?

Looking forward to an update that would address this. Or in case it's already possible I would be happy with a how-to guide. Thanks!

**Reply by brandonkelly** [reply]
The [`relatedToEntries`](https://craftcms.com/docs/4.x/graphql.html#relations) (et al.) params will search both ways (forward and reverse), which should generally work for you unless you have two sections that each have different Entries fields pointing at each other.

**Reply by ctigelaar** [reply]
Thanks for your quick reply Brandon. I know I can narrow down my entries based on `relatedToEntries`, but that is not what I need.

1. Let's say I have two sections `organisations` and `locations`. In an organisation I have a `organisationLocations` entries selection field, which allows selection from the `locations` section. 
2. I want a list of all locations based on a search in `organisations`. So I use an `entries()` on section `locations` with a `relatedToEntries` argument with a `search` in `organisations`. So far so good, I'm getting the list of `locations`.
3. But the `entries()` is returning me results of `locations`, that's correct, but there is no way to get the reverse related entries from `organisations` here, is there?
4. That's the key piece I'm missing. I hear you thinking: "you can query with `entries()` for `organisations` initially, then you have all the data you need." But that would give me a list of `organisations` with their locations in the `organisationLocations` field, instead of a list of `locations` with their organisations. It matters, because it also affects pagination. I need to paginate on the `locations`, not on the `organisations`.

I hope I made usecase more clear. It seems like all ingredients are available in the CMS/database. And maybe I'm still overlooking some built-in functionality here, but please let me know then. Thanks.

**Reply by brandonkelly** [reply]
Yeah currently that would require an additional query for each resulting location, which I understand isn’t ideal.

> It seems like all ingredients are available in the CMS/database.

Not really, which is the reason this isn’t resolved yet. Craft doesn’t have any way to reference/eager-load reverse relations currently. See #8424 for the discussion on that.
